freezefrog/__init__.py

```diff
@@ -28,100 +65,124 @@ def __instancecheck__(self, obj):
     """
-    A datetime replacement that lets you set utcnow() using set_utcnow().
+    A `datetime` mock class.
 
-    The clock starts ticking after calling set_utcnow().
+    Set the start datetime and the system timezone with `start(dt, tz)`. The
+    clock starts ticking after calling `start`.
     """
-    tz_delta = None
 
-    @classmethod
-    def utcnow(cls, *args, **kwargs):
-        if not hasattr(cls, 'dt'):
-            raise NotImplementedError(
-                'use {}.set_utcnow(datetime) first'.format(cls.__name__)
-            )
-        return cls._utcnow()
+    dt = None
+    tz = None
+    _start = None
 
     @classmethod
-    def set_utcnow(cls, dt):
+    def start(cls, dt, tz, tick, fold):
         cls.dt = dt
+        cls.tz = tz
+        cls._start = time.monotonic() if tick else None
+
+        if is_pytz(tz):
+            if not PYTZ_SUPPORT:
+                raise Exception(
+                    "pytz not supported, please install pytz first"
+                )
+
+            cls.is_dst = get_pytz_dst_from_fold(dt, tz, fold)
+            cls.now = cls.now_with_pytz
+            cls.utcnow = cls.utcnow_with_pytz
+        else:
+            cls.fold = fold
+            cls.now = cls.now_with_datetime_tz
+            cls.utcnow = cls.utcnow_with_datetime_tz
 
     @classmethod
-    def set_tz_delta(cls, tz_delta):
-        cls.tz_delta = tz_delta
+    def time_since_start(cls):
+        if cls._start is None:
+            return datetime.timedelta(seconds=0)
+        return datetime.timedelta(seconds=time.monotonic() - cls._start)
 
     @classmethod
-    def _utcnow(cls):
-        if not hasattr(cls, '_start'):
-            cls._start = real_datetime.utcnow()
-        return (real_datetime.utcnow() - cls._start) + cls.dt
+    def now_with_datetime_tz(cls, tz=None):
+        if tz is None:
+            return cls.dt + cls.time_since_start()
+        return (
+            cls.dt.replace(tzinfo=cls.tz, fold=cls.fold).astimezone(
+                datetime.timezone.utc
+            )
+            + cls.time_since_start()
+        ).astimezone(tz)
 
     @classmethod
-    def now(cls, *args, **kwargs):
-        if cls.tz_delta is not None:
-            return cls.utcnow() + cls.tz_delta
-        else:
-            raise Exception(
-                '{}.now() requires setting tz_delta'.format(cls.__name__)
-            )
+    def now_with_pytz(cls, tz=None):
+        if tz is None:
+            return cls.dt + cls.time_since_start()
+        return tz.normalize(
+            pytz.UTC.normalize(cls.tz.localize(cls.dt, is_dst=cls.is_dst))
+            + cls.time_since_start()
+        )
+
+    @classmethod
+    def today(cls):
+        return cls.now()
 
+    @classmethod
+    def utcnow_with_datetime_tz(cls):
+        return cls.now(tz=datetime.timezone.utc).replace(tzinfo=None)
 
-class FakeFixedDateTime(FakeDateTime):
     @classmethod
-    def _utcnow(cls):
-        return cls.dt
+    def utcnow_with_pytz(cls):
```

Comment:
Any reason we need this?

Reply:
To support both `pytz` and `datetime`.

Reply:
Why would you need this though? You just need to return the naive time in UTC. You can just always the `datetime` implementation, or am I missing something?

Reply:
Regardless of whether we pass a pytz or a native timezone to the `FreezeTime` constrctur, the mocked `now()` should work with either a `pytz` or a native timezone, like it would normally.

Reply:
`now()` accepts `datetime` and `pytz` timezone objects, and `utcnow()` uses the `datetime` implementation. Done in https://github.com/closeio/freezefrog/pull/13/commits/a16ee262432487511d0ea8b831218d33450deac5.